File: 송현진/백준_JAVA/20922_겹치는건싫어.java
import java.io.*;
import java.util.*;

class Solution {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] arr = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < arr.length; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int ans = 0;
        int start = 0;
        int end = 0;
        // 100000이하의 정수로 이루어진 ㅇㅇ
        int[] cnt = new int[1000001];

        while (end < arr.length) {
            while (end < arr.length && cnt[arr[end]] + 1 <= M) {
                cnt[arr[end]]++;
                end++;
            }
            int len = end - start;
            ans = Math.max(len, ans);
            cnt[arr[start]]--;
            start++;
        }
        System.out.println(ans);
    }
}

File: 송현진/백준_KOTLIN/20922_겹치는건싫어.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() {
    var br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())

    var N = st.nextToken().toInt()
    var num = st.nextToken().toInt()

    st = StringTokenizer(br.readLine())

    var arr = Array(N){0}
    for(i in 0 until N){
        arr[i] = st.nextToken().toInt()
    }

    var ans = 0
    var start = 0
    var end = 0
    var cnt = Array(100001){0}

    while(end < arr.size){
        //최장 연속 부분 수열 && 같은 숫자 num보다 작거나 같은 개수만 가지기
        //cnt[arr[end]] + 1 == +1 더한수가 정해진 num을 넘는지 확인
        while(end < arr.size && cnt[arr[end]] + 1 <= num){
            cnt[arr[end]]++
            end++
        }
        var len = end-start
        ans = Math.max(len,ans)
        //앞에서 하나씩 지우면서 start옮기기 -> 앞에서 하나씩 제거되면서 겹치던 숫자가 사라질 수 있음
        cnt[arr[start]]--
        start++

//        for(i in cnt.indices){
//            print("${cnt[i]} ")
//        }
//        println()
    }
    println(ans)
}